### PR TITLE
KAFKA-17154: New consumer subscribe may join group without a call to consumer.poll

### PR DESCRIPTION
**Pull request by FrankYang0529**
To fulfill "rebalances will only occur during an active call to KafkaConsumer#poll(Duration)", we should not send `JoinRequest` after `AsyncKafkaConsumer#subscribe`. Add a flag `subscriptionUpdated` to `AbstractMembershipManager#onSubscriptionUpdated`. When calling `AsyncKafkaConsumer#subscribe`, set the flag to true. When calling `AsyncKafkaConsumer#poll`, send the `JoinRequest` if the flag is true.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by TaiJuWu** [comment]
Hi @FrankYang0529 , I opened a PR for this behavior.
Could you take a look?  Thanks a lot!  

https://github.com/apache/kafka/pull/17175

**Comment by FrankYang0529** [comment]
> I opened a PR for this behavior. Could you take a look? Thanks a lot!
> 
> #17175

Hi @TaiJuWu, thanks for the PR. I also have a similar test case `testPollSendRequestForRebalance`. We can also have your test after this PR is merged.

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java
Comment (TaiJuWu) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:3457]: Maybe we need to make `application thread`sleep here to make sure `network thread` does not send join request.
WDYT?

**Comment by TaiJuWu** [comment]
> > I opened a PR for this behavior. Could you take a look? Thanks a lot!
> > #17175
> 
> Hi @TaiJuWu, thanks for the PR. I also have a similar test case `testPollSendRequestForRebalance`. We can also have your test after this PR is merged.

Hi @FrankYang0529 , I will close my PR because they do same thing and I leave a comment for `testPollSendRequestForRebalance`.

**Comment by lianetm** [comment]
Hey @FrankYang0529, so seems this draft is the PR we're keeping right? Let me know when it's ready for review and I'll be happy to take a look. Thanks!

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java
Comment (FrankYang0529) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:3457]: Hi @TaiJuWu, `application` and `network` are different threads. I'm not sure why making `application` thread sleep can let `network` thread doesn't send join request. Also , this test would like to make sure `KafkaConsumer#subscribe` doesn't send join request, so we should not have other operations to make the result. Could you elaborate more on your idea? Thank you.

**Comment by FrankYang0529** [comment]
Hi @lianetm, yes, the PR is ready for review. Thank you.

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java
Comment (TaiJuWu) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:3457]: Sorry for the confusion.

The reason for making the application sleep here is to give the network enough time to send the join request even without poll. This helps to reproduce the bug more easily and consistently.

Additionally, to ensure the behavior we’ve promised in the documentation, the two operations (subscribe and poll) should not be too close together in the timeline in this test. Does that make sense?

Comment (FrankYang0529) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:3457]: Hi @TaiJuWu, thanks for the suggestion. It's hard to say waiting how much time can make sure `KafkaConsumer#subscribe` doesn't send a join request. I add two unit tests to check the behavior.

In `ApplicationEventProcessorTest#testSubscriptionChangeEvent`, we check `ApplicationEventProcessor#process(SubscriptionChangeEvent)` doesn't call `ConsumerMembershipManager#maybeJoinGroup`.
In `ConsumerMembershipManagerTest#testOnSubscriptionUpdatedTransitionsToJoiningOnlyIfNotInGroup`, we check `ConsumerMembershipManager#onSubscriptionUpdated` keeps member state as `UNSUBSCRIBED` and calling `transitionToJoining` in `maybeJoinGroup`.

Comment (TaiJuWu) [reply clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:3457]: Thanks for your explanation.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerMembershipManagerTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerMembershipManagerTest.java]: these 2 lines are a bit confusing (we trigger "onSubscriptionUpdated" and expect "subscriptionUpdated" to be false). I would say that the boolean var name is what we should review, because it truly indicates if the member should join (related to comment on the membershipManager.subscriptionUpdated var)

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerMembershipManagerTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerMembershipManagerTest.java]: what about `testSubscriptionUpdateDoesNotTransitionToJoining` (or similar)? (not transitioning to joining is really the core bit we were missing and are fixing/test here..and is actually consistent with the name you have for the similar test right below)

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerMembershipManagerTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerMembershipManagerTest.java:1633]: should we verify here that there was no `transitionToJoining`? (to be sure that the transition is only triggered after the call to `maybeJoinGroup`)

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: typo testPoll**Sends**RequestForRebalance?  (and maybe clearer `testPollSendsRequestToJoin`?)

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractMembershipManager.java
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractMembershipManager.java]: this is truly more to know if we should join after the subscription has been updated right? (true if not in group)

**Review by lianetm — COMMENTED** [review]
Hey @FrankYang0529 , thanks for the patch! Left some comments for consideration.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerMembershipManagerTest.java
Comment (FrankYang0529) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerMembershipManagerTest.java]: Change variable from `subscriptionUpdated` to `shouldTransitionToJoining` and add more comments for it. Thank you.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerMembershipManagerTest.java
Comment (FrankYang0529) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerMembershipManagerTest.java]: Thanks for the suggestion. I think whether the member is in the group is important as well, because it determines whether to change `shouldTransitionToJoining` variable. I update the test case name to `testOnSubscriptionUpdatedDoesNotSetShouldTransitionToJoiningIfInGroup` and another test case below to `testOnSubscriptionUpdatedSetShouldTransitionToJoiningIfNotInGroup`. Do you think it makes sense? Thank you.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerMembershipManagerTest.java
Comment (FrankYang0529) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerMembershipManagerTest.java:1633]: Yes, added it. Thanks.

**Comment by FrankYang0529** [comment]
Hi @lianetm, I found that `PlaintextConsumerSubscriptionTest#testSubscribeInvalidTopic` is flaky in this PR, so I will convert to draft first. Thanks.

**Comment by lianetm** [comment]
Hey @FrankYang0529, just in case it helps, we have this https://issues.apache.org/jira/browse/KAFKA-17286 related to that flaky test with some ideas in the comments.

**Comment by FrankYang0529** [comment]
> Hey @FrankYang0529, just in case it helps, we have this https://issues.apache.org/jira/browse/KAFKA-17286 related to that flaky test with some ideas in the comments.

Yes, that helps. I found that both in trunk and this PR, the failed result is reproducible by adding `Thread.sleep(1000)` at the end of `PlaintextConsumerSubscriptionTest#testSubscribeInvalidTopic`.

Do you think we let https://issues.apache.org/jira/browse/KAFKA-17286 handle this flaky? Or we should solve it in this PR? Thanks.

**Comment by lianetm** [comment]
Hey @FrankYang0529 , sorry for the late reply, I was traveling last week but I'm back, so will be taking another look at this. You can mark it as ready, we can leave the fix for the flaky test separate given that we already understand the root cause not related to this change. I'll take another look at this one between today and tomorrow. Thanks!

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractMembershipManager.java
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractMembershipManager.java]: I wonder if this check is still needed here? It made sense before because we were doing the actual join, but now we're just setting a flag (that looks more like a "subscriptionUpdated" var now). Then on the `maybeJoinGroup` is where we need to check that the state if `UNSUBSCRIBED` and `subscriptionUpdated` then join. Makes sense?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessor.java
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessor.java]: lets remove this please

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessor.java
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessor.java]: ditto

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerSubscriptionTest.scala
Comment (lianetm) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerSubscriptionTest.scala]: remove?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractMembershipManager.java
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractMembershipManager.java]: I don't quite get how this change relates to this PR, why is it that we need this here?

Also, this is effectively ignoring the leave group (unsubscribe) if JOINING, which I would expect is not right (the member will remain JOINING and may even become STABLE, never leave). Before this change, JOINING + consumer.unsubscribe => LEAVING (member would run the full leave flow and attempt to send the leave HB). With this change, JOINING + consumer.unsubscribe => still JOINING. Is that the intention?

I know that the leave while joining has challenges to solve so that it can be processed correctly (KIP-1082), but I would expect that we keep the intention we had here, even after the KIP-1082 fixes: we should attempt to leave the group when there's a call to unsubscribe while waiting for the join response (JOINING) because the broker may have already processed the join.

**Review by lianetm — COMMENTED** [review]
Hey @FrankYang0529, I took another full pass and left some comments for consideration. Thanks!

**Comment by FrankYang0529** [comment]
Hi @lianetm, I'm sorry that there was a testing commit. I've removed it. Also, I change `shouldTransitionToJoining` to `hasSubscriptionUpdated` and set `hasSubscriptionUpdated` to true in `AbstractMembershipManager#hasSubscriptionUpdated` without checking subscription state.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractMembershipManager.java
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractMembershipManager.java]: nit: what about simply `subscriptionUpdated` ...to track whether the subscription has been updated

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerMembershipManagerTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerMembershipManagerTest.java]: there is no "shouldTransitionToJoining" var/concept anymore, so should we update this ? maybe simply `...DoesNotTransitionToJoiningIfInGroup`

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerMembershipManagerTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerMembershipManagerTest.java]: `testOnSubscriptionUpdatedTransitionsToJoiningIOnPollfNotInGroup`?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ShareMembershipManagerTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/ShareMembershipManagerTest.java]: let's make sure to align these with the names in the consumer tests if they change with the comments above

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessorTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessorTest.java]: I would say this test only makes sense withGroupId=true right? if there is no groupId, the processor will have null HBMgr, MembershipMgr (there cannot be a `SubscriptionChange` event without a groupId, the api call to `subscribe` requires groupId)

**Review by lianetm — COMMENTED** [review]
Hey @FrankYang0529 , thanks for the updates. Just some minor comments left.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractMembershipManager.java
Comment (FrankYang0529) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractMembershipManager.java]: Yes, update the variable and function to `subscriptionUpdated`. Thank you.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerMembershipManagerTest.java
Comment (FrankYang0529) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerMembershipManagerTest.java]: Updated the test case name to `testOnSubscriptionUpdatedDoesNotTransitionToJoiningIfInGroup`.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerMembershipManagerTest.java
Comment (FrankYang0529) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerMembershipManagerTest.java]: Changed the test case name to `testOnSubscriptionUpdatedTransitionsToJoiningOnPollIfNotInGroup`. Thanks.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ShareMembershipManagerTest.java
Comment (FrankYang0529) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/ShareMembershipManagerTest.java]: Yes, align test case names between `ConsumerMembershipManagerTest` and `ShareMembershipManagerTest`.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessorTest.java
Comment (FrankYang0529) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessorTest.java]: Yes, I leave `withGroupId=true` case here. Thank you.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractMembershipManager.java
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractMembershipManager.java]: nit: what do you think about calling this something like `onConsumerPoll`? Just to make the func self explanatory (on poll -> transition to joining), no need to find usages to understand what this is all about. No strong feeling though, up to you.

**Review by lianetm — COMMENTED** [review]
Hey @FrankYang0529 , could you please update the PR description where it says:
> When calling AsyncKafkaConsumer#subscribe, send the JoinRequest if the flag is true.

I guess you meant when calling poll transition to joining if the subscription has been updated?

With that (and the last nit up to you) this looks good to me. Thanks!

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractMembershipManager.java
Comment (FrankYang0529) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractMembershipManager.java]: Thanks for the suggestion. I updated PR description and change `maybeJoinGroup` to `onConsumerPoll`.

**Review by lianetm — APPROVED** [review]
Thanks @FrankYang0529! LGTM.

**Comment by lianetm** [comment]
FYI, I filed https://issues.apache.org/jira/browse/KAFKA-17623 for a flaky consumer integration test that I noticed here, has been flaky for a while.

**Comment by FrankYang0529** [comment]
> FYI, I filed https://issues.apache.org/jira/browse/KAFKA-17623 for a flaky consumer integration test that I noticed here, has been flaky for a while.

Okay. I will take a look. Thanks for filing the issue.

**Comment by chia7712** [comment]
This PR causes the following build error:
```
/home/chia7712/project/kafka/clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:3500: error: method requestGenerated(MockClient,ApiKeys) is already defined in class KafkaConsumerTest
    private boolean requestGenerated(MockClient client, ApiKeys apiKey) {
```

**Comment by lianetm** [comment]
uhm builds completed successfully but I guess it conflicted with a recent merge. Checking now and will open PR right away. Thanks @chia7712 !

**Comment by lianetm** [comment]
Removing dup method here https://github.com/apache/kafka/pull/17291

**Comment by mumrah** [comment]
Once ASF signs off on the merge queue, we can completely avoid this class of failure on trunk. According to Infra team, we are a month or two away from being able to enable it.

**Comment by lianetm** [comment]
That will be helpful, it was really my miss not asking to have this rebased one last time.

**Pull request merged**
